### PR TITLE
EC2 multi-node environment - revised

### PR DESCRIPTION
**Pull request by johnkim76**

### CODE REVIEW

**Comment by johnkim76** [comment]
[(13) Update ec2-multi in catasb PR to address concerns](https://trello.com/c/2pk2LZaK/609-13-update-ec2-multi-in-catasb-pr-to-address-concerns)

File: ansible/roles/env_hacks/tasks/main.yml [hunk 1]
```diff
@@ -1,13 +1,8 @@
...
-  # startup.
-  #
-  # We've seen some permission errors testing certain containers with PVs
-  # May be related to this issue:
-  #   https://github.com/openshift/origin/issues/13496https://github.com/openshift/origin/issues/13496
-  # Putting in a workaround for manually opening up the permissions
+- include: minimal.yml
+  when: minimal
 
+<<<<<<< HEAD
```
Comment (djwhatle) [inline ansible/roles/env_hacks/tasks/main.yml]: Looks like there are a few artifacts from rebasing/merging here.

File: ansible/roles/env_hacks/tasks/main.yml [hunk 2]
```diff
@@ -61,3 +56,7 @@
         when: openshift_client_version == '3.7' or openshift_client_version == 'latest'
       when: coalmine
     when: coalmine is defined
+=======
+- include: multi_node.yml
+  when: multi_node
+>>>>>>> revised update for the ec2 multi-node environment
```
Comment (djwhatle) [inline ansible/roles/env_hacks/tasks/main.yml]: Here as well

File: ansible/roles/env_hacks/tasks/main.yml [hunk 3]
```diff
@@ -61,3 +56,7 @@
         when: openshift_client_version == '3.7' or openshift_client_version == 'latest'
       when: coalmine
     when: coalmine is defined
+=======
```
Comment (djwhatle) [inline ansible/roles/env_hacks/tasks/main.yml]: And here.

**Comment by djwhatle** [comment]
I'm still looking, just wanted to submit initial errors ran into.

**Comment by djwhatle** [comment]
Let's not forget to address the issue of assuming the user is set up with passwordless sudo on the ansible control node. Can simulate this by removing passwordless sudo from local user account and seeing where it gets stuck.

**Comment by djwhatle** [comment]
We are repeatedly running into this error on the "Apply the SDN Config" step:

```
fatal: [ec2-54-198-135-0.compute-1.amazonaws.com]: FAILED! => {"changed": true
 "cmd": "oc apply -f /tmp/console-ansible-79l9DD"
 "delta": "0:00:00.208964"
 "end": "2018-03-15 00:06:20.779197"
 "msg": "non-zero return code"
 "rc": 1
 "start": "2018-03-15 00:06:20.570233"
 "stderr": "unable to recognize \"/tmp/console-ansible-79l9DD/sdn-ovs.yaml\": no matches for apps/
 Kind=DaemonSet\nunable to recognize \"/tmp/console-ansible-79l9DD/sdn.yaml\": no matches for apps/
 Kind=DaemonSet"
 "stderr_lines": ["unable to recognize \"/tmp/console-ansible-79l9DD/sdn-ovs.yaml\": no matches for apps/
 Kind=DaemonSet"
 "unable to recognize \"/tmp/console-ansible-79l9DD/sdn.yaml\": no matches for apps/
 Kind=DaemonSet"]
 "stdout": "imagestreamtag \"node:v3.9\" created\nserviceaccount \"sdn\" created\nclusterrolebinding \"sdn-cluster-reader\" created\nclusterrolebinding \"sdn-reader\" created"
 "stdout_lines": ["imagestreamtag \"node:v3.9\" created"
 "serviceaccount \"sdn\" created"
 "clusterrolebinding \"sdn-cluster-reader\" created"
 "clusterrolebinding \"sdn-reader\" created"]}
```

**Comment by djwhatle** [comment]
I noticed that EC2 instances are spawned one at a time in this flow. I think it would be ideal to have all instances come up at the same time, I'll see if I can figure out how to do this.

Also, I'm hitting an issue where a similarly name tagged EC2 instance from a non multi-node deploy is getting pulled into the inventory. I'd propose that we move to using unique AWS Instance ID's for tracking in the future. This would also help us avoid the issue with getting AWS instances from multiple regions pulled into provision automation by mistake.

File: ansible/roles/aws_infrastructure/templates/openshift_ansible_inventory_ose.j2 [hunk 4]
```diff
@@ -0,0 +1,90 @@
+[OSEv3:children]
+masters
+etcd
+nodes
+
+[OSEv3:vars]
+ansible_become=yes
+ansible_ssh_user={{ aws_username }}
+
```
[the remark this answers is not in the document]
Comment (djwhatle) [reply ansible/roles/aws_infrastructure/templates/openshift_ansible_inventory_ose.j2]: Went ahead and added this.

**Comment by djwhatle** [comment]
Tested Origin and Enterprise successfully.

**For both Origin and Enterprise**
`~/.bashrc` contained

```
export AWS_ACCESS_KEY_ID="..."                                                                                                                                                                                                           
export AWS_SECRET_ACCESS_KEY="..."                                                                                                                                                                                   
export TARGET_DNS_ZONE="aws.dog8code.com"                                                                                                                                                                                                                 
export AWS_REGION="us-east-1"                                                                                                                                                                                                                             
export AWS_SSH_KEY_NAME="sample-aws-key-name"                                                                                                                                                                                                                    
export AWS_SSH_PRIV_KEY_PATH="/example/ssh-key.pem" 
```
**For Origin:**
`config/my_vars.yml` contained
```
openshift_ansible_repo_branch: 'release-3.9'   
deploy_origin: true
```

**For Enterprise:**
`config/my_vars.yml` contained
```
 rhn_user: ***
 rhn_passwd: ***
 aos_pool_id: ***
```

Planning on waiting until 3.9 fixes are merged into catasb to run regression tests against other env types.

**Comment by djwhatle** [comment]
Error trying to run local/linux against latest during `oc cluster up`: 

```
Error: unknown flag: --service-catalog
```

```
TASK [openshift_setup : Run oc cluster up to start the cluster] ***********************************************************************************************************************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "/home/dwhatley/bin/oc cluster down && /home/dwhatley/bin/oc cluster up --routing-suffix=172.17.0.1.nip.io --image=docker.io/openshift/origin --tag=latest --base-dir=/tmp/openshift.local.clusterup --service-catalog=true", "delta": "0:00:00.318107", "end": "2018-04-10 15:15:22.673424", "msg": "non-zero return code", "rc": 1, "start": "2018-04-10 15:15:22.355317", "stderr": "Error: unknown flag: --service-catalog\n\n\nUsage:\n  oc cluster up [flags]\n\nExamples:\n  # Start OpenShift using a specific public host name\n  oc cluster up --public-hostname=my.address.example.com\n  \n  # Specify which set of image streams to use\n  oc cluster up --image-streams=centos7\n\nOptions:\n      --base-dir='': Directory on Docker host for cluster up configuration\n      --enable=[*]: A list of components to enable.  '*' enables all on-by-default components, 'foo' enables the component named 'foo', '-foo' disables the component named 'foo'.\nAll components: registry, router, service-catalog, template-service-broker, web-console\nDisabled-by-default components: service-catalog, template-service-broker\n  -e, --env=[]: Specify a key-value pair for an environment variable to set on OpenShift container\n      --forward-ports=false: Use Docker port-forwarding to communicate with origin container. Requires 'socat' locally.\n      --http-proxy='': HTTP proxy to use for master and builds\n      --https-proxy='': HTTPS proxy to use for master and builds\n      --image='openshift/origin': Specify the images to use for OpenShift\n      --image-streams='centos7': Specify which image streams to use, centos7|rhel7\n      --no-proxy=[]: List of hosts or subnets for which a proxy should not be used\n      --public-hostname='': Public hostname for OpenShift cluster\n      --routing-suffix='': Default suffix for server routes\n      --server-loglevel=0: Log level for OpenShift server\n      --skip-registry-check=false: Skip Docker daemon registry check\n      --use-existing-config=false: Use existing configuration if present\n      --write-config=false: Write the configuration files into host config dir\n\nUse \"oc options\" for a list of global command-line options (applies to all commands).", "stderr_lines": ["Error: unknown flag: --service-catalog", "", "", "Usage:", "  oc cluster up [flags]", "", "Examples:", "  # Start OpenShift using a specific public host name", "  oc cluster up --public-hostname=my.address.example.com", "  ", "  # Specify which set of image streams to use", "  oc cluster up --image-streams=centos7", "", "Options:", "      --base-dir='': Directory on Docker host for cluster up configuration", "      --enable=[*]: A list of components to enable.  '*' enables all on-by-default components, 'foo' enables the component named 'foo', '-foo' disables the component named 'foo'.", "All components: registry, router, service-catalog, template-service-broker, web-console", "Disabled-by-default components: service-catalog, template-service-broker", "  -e, --env=[]: Specify a key-value pair for an environment variable to set on OpenShift container", "      --forward-ports=false: Use Docker port-forwarding to communicate with origin container. Requires 'socat' locally.", "      --http-proxy='': HTTP proxy to use for master and builds", "      --https-proxy='': HTTPS proxy to use for master and builds", "      --image='openshift/origin': Specify the images to use for OpenShift", "      --image-streams='centos7': Specify which image streams to use, centos7|rhel7", "      --no-proxy=[]: List of hosts or subnets for which a proxy should not be used", "      --public-hostname='': Public hostname for OpenShift cluster", "      --routing-suffix='': Default suffix for server routes", "      --server-loglevel=0: Log level for OpenShift server", "      --skip-registry-check=false: Skip Docker daemon registry check", "      --use-existing-config=false: Use existing configuration if present", "      --write-config=false: Write the configuration files into host config dir", "", "Use \"oc options\" for a list of global command-line options (applies to all commands)."], "stdout": "", "stdout_lines": []}
```

**Comment by djwhatle** [comment]
Trying to run this against Origin currently results in failure on all nodes at this step:
```
TASK [openshift_node : Install Node package, sdn-ovs, conntrack packages] 
failed: [] (item={u'name': u'origin-node-3.7.1'}) => {"atte
mpts": 2, "changed": false, "item": {"name": "origin-node-3.7.1"}, "msg": "Failure talking to yum: Co
uld not contact any CDS load balancers: rhui2-cds01.us-east-1.aws.ce.redhat.com, rhui2-cds02.us-east-
1.aws.ce.redhat.com."}
```

The issue here appears to be related to DNS. Specifically, the created systems are pointing at themselves for DNS resolution and then failing to resolve hostnames.

I was curious if the issue was caused by this commit: https://github.com/openshift/openshift-ansible/commit/f38aaacd53858d288d268645bca429210cf30703

After removing the changes in that commit, the problem persisted. I plan to continue looking into whether dnsmasq may be misconfigured.

**Comment by djwhatle** [comment]
^ The issue mentioned above with etcd was due to UDP ports not being opened between nodes like they should have been. 

Fixed by https://github.com/fusor/catasb/pull/238

**Pull request closed without merging**